# 1. Import Libraries

In [55]:
import pandas as pd

In [56]:
import numpy as np

# 2. Load Dataset

In [57]:
df = pd.read_csv("Messy_Company_Employee_Dataset.csv")
df

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,NaN,Data Scientist,48.0,Other,108371.0,13.2,2017-06-13,Pune,5.0,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.1,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28.0,Female,119400.0,13.2,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,NaN,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.9,2018-07-24,Hyderabad,4.0,Hybrid
...,...,...,...,...,...,...,...,...,...,...,...,...
152,EMP0132,Manya Rao,Marketing,Content Strategist,30.0,Female,49522.0,0.6,2021-11-25,Delhi,2.0,Remote
153,EMP0073,Karan Kapoor,Customer Success,Customer Success Manager,42.0,MALE,43331.0,1.1,2018-03-23,Jaipur,3.0,Hybrid
154,EMP0124,Saira Kapoor,Engineering,QA Engineer,33.0,Male,94812.0,7.2,2024-06-12,Jaipur,4.0,Hybrid
155,EMP0020,Sara Ali,Sales,Sales Executive,22.0,Female,51745.0,2.4,2020-11-19,NaN,2.0,Office


In [58]:
df.head()

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,NaN,Data Scientist,48.0,Other,108371.0,13.2,2017-06-13,Pune,5.0,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.1,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28.0,Female,119400.0,13.2,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,NaN,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.9,2018-07-24,Hyderabad,4.0,Hybrid


# 3.Dataset Inspection

In [59]:
df.shape

(157, 12)

In [60]:
df.columns

Index(['Employee_ID', 'Employee_Name', 'Department', 'Job_Title', 'Age',
       'Gender', 'Annual_Salary', 'Experience_Years', 'Joining_Date', 'City',
       'Performance_Score', 'Work_Mode'],
      dtype='object')

In [61]:
df.dtypes

,0
Employee_ID,object
Employee_Name,object
Department,object
Job_Title,object
Age,float64
Gender,object
Annual_Salary,float64
Experience_Years,float64
Joining_Date,object
City,object


In [62]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157 entries, 0 to 156
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Employee_ID        157 non-null    object 
 1   Employee_Name      157 non-null    object 
 2   Department         152 non-null    object 
 3   Job_Title          157 non-null    object 
 4   Age                153 non-null    float64
 5   Gender             152 non-null    object 
 6   Annual_Salary      152 non-null    float64
 7   Experience_Years   154 non-null    float64
 8   Joining_Date       157 non-null    object 
 9   City               152 non-null    object 
 10  Performance_Score  154 non-null    float64
 11  Work_Mode          155 non-null    object 
dtypes: float64(4), object(8)
memory usage: 14.8+ KB


In [63]:
df.describe()

,Age,Annual_Salary,Experience_Years,Performance_Score
count,153.000000,152.000000,154.000000,154.000000
mean,38.130719,89240.592105,8.990909,3.525974
std,9.361661,22626.172309,4.983874,0.930233
min,21.000000,43331.000000,0.600000,2.000000
25%,30.000000,71198.500000,4.450000,3.000000
50%,39.000000,90046.500000,9.350000,4.000000
75%,46.000000,107328.000000,13.200000,4.000000
max,52.000000,162942.000000,17.400000,5.000000


In [64]:
df.describe(include="object")

,Employee_ID,Employee_Name,Department,Job_Title,Gender,Joining_Date,City,Work_Mode
count,157,157,152,157,152,157,152,155
unique,150,130,10,23,5,147,13,5
top,EMP0075,Nikhil Das,Marketing,Data Scientist,Male,2025-08-17,Chandigarh,Remote
freq,2,3,23,13,75,2,24,56


# 4.Check Missing Values

In [65]:
df.isnull().sum()

,0
Employee_ID,0
Employee_Name,0
Department,5
Job_Title,0
Age,4
Gender,5
Annual_Salary,5
Experience_Years,3
Joining_Date,0
City,5


In [66]:
df.isnull().sum().sum()

np.int64(32)

# 5. Identify Duplicate Records

In [67]:
df.duplicated().sum()

np.int64(7)

In [68]:
df[df.duplicated()]

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
55,EMP0097,Zoya Shah,Finance,Financial Analyst,41.0,Male,81531.0,4.1,2023-06-05,Mumbai,4.0,Office
61,EMP0075,Manya Joshi,Sales,Sales Manager,29.0,Male,92623.0,11.9,2025-08-17,Jaipur,4.0,Hybrid
100,EMP0139,Simran Sharma,Finance,Accountant,32.0,Female,110617.0,10.3,2021-08-03,Mumbai,3.0,Office
104,EMP0021,Riya Khan,Operations,Operations Executive,49.0,Female,71549.0,4.6,2024-09-20,Chandigarh,3.0,Hybrid
110,EMP0048,Nikhil Dar,Finance,Financial Analyst,27.0,Female,NaN,10.1,2020-06-18,Chandigarh,2.0,Office
139,EMP0150,Nikhil Khan,Finance,Finance Manager,43.0,Male,129905.0,16.3,2018-04-10,Chennai,3.0,Remote
156,EMP0120,Saira Sheikh,Marketing,Marketing Executive,52.0,Male,107328.0,17.4,2020-11-25,Bengaluru,2.0,Remote


In [69]:
df[df.duplicated(keep=False)].sort_values("Employee_ID")

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
70,EMP0021,Riya Khan,Operations,Operations Executive,49.0,Female,71549.0,4.6,2024-09-20,Chandigarh,3.0,Hybrid
104,EMP0021,Riya Khan,Operations,Operations Executive,49.0,Female,71549.0,4.6,2024-09-20,Chandigarh,3.0,Hybrid
78,EMP0048,Nikhil Dar,Finance,Financial Analyst,27.0,Female,NaN,10.1,2020-06-18,Chandigarh,2.0,Office
110,EMP0048,Nikhil Dar,Finance,Financial Analyst,27.0,Female,NaN,10.1,2020-06-18,Chandigarh,2.0,Office
11,EMP0075,Manya Joshi,Sales,Sales Manager,29.0,Male,92623.0,11.9,2025-08-17,Jaipur,4.0,Hybrid
61,EMP0075,Manya Joshi,Sales,Sales Manager,29.0,Male,92623.0,11.9,2025-08-17,Jaipur,4.0,Hybrid
25,EMP0097,Zoya Shah,Finance,Financial Analyst,41.0,Male,81531.0,4.1,2023-06-05,Mumbai,4.0,Office
55,EMP0097,Zoya Shah,Finance,Financial Analyst,41.0,Male,81531.0,4.1,2023-06-05,Mumbai,4.0,Office
92,EMP0120,Saira Sheikh,Marketing,Marketing Executive,52.0,Male,107328.0,17.4,2020-11-25,Bengaluru,2.0,Remote
156,EMP0120,Saira Sheikh,Marketing,Marketing Executive,52.0,Male,107328.0,17.4,2020-11-25,Bengaluru,2.0,Remote


In [70]:
df["Employee_ID"].duplicated().sum()

np.int64(7)

In [71]:
df = df.drop_duplicates(subset=["Employee_ID"], keep="last")

In [72]:
df.duplicated().sum()

np.int64(0)

# 6. Checking Inconsistent Categorical Values

In [73]:
# Department
df["Department"].value_counts(dropna = False)

,count
Department,
Marketing,22
Engineering,21
Customer Success,20
Sales,18
Human Resources,16
Operations,16
Data & Analytics,15
Finance,15
NaN,5


In [74]:
# Gender
df["Gender"].value_counts(dropna = False)

,count
Gender,
Male,71
Female,65
Other,7
NaN,5
female,1
MALE,1


In [75]:
# City
df["City"].value_counts(dropna = False)

,count
City,
Pune,23
Chandigarh,22
Jaipur,17
Hyderabad,14
Srinagar,13
Mumbai,13
Delhi,11
Bengaluru,11
Chennai,9


In [76]:
# Work_Mode
df["Work_Mode"].value_counts(dropna=False)

,count
Work_Mode,
Remote,54
Hybrid,48
Office,44
NaN,2
REMOTE,1
remote,1


In [77]:
df.head(2)

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,NaN,Data Scientist,48.0,Other,108371.0,13.2,2017-06-13,Pune,5.0,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.1,2021-09-15,Mumbai,2.0,Office


# 7.Remove Unwanted Spaces

In [ ]:
object_columns = df.select_dtypes(include="object").columns
for col in object_columns:
  df[col] = df[col].str.strip()

In [81]:
df["City"].unique()

array(['Pune', 'Mumbai', 'Bengaluru', 'Jaipur', 'Hyderabad', 'Chennai',
       'Chandigarh', 'Srinagar', 'Kolkata', 'delhi', nan, 'Delhi',
       'DELHI'], dtype=object)

# 8.Standardize Categorical Entries

In [ ]:
# for Department
df["Department"] = df["Department"].str.title()

In [ ]:
# for Gender
df["Gender"] = df["Gender"].str.title()

In [ ]:
# for City
df["City"] = df["City"].str.title()

In [ ]:
# for Work Mode
df["Work_Mode"] = df["Work_Mode"].str.title()

In [86]:
# Let's Check for Department
df["Department"].value_counts(dropna = False)

,count
Department,
Engineering,23
Marketing,22
Customer Success,20
Sales,18
Human Resources,16
Operations,16
Data & Analytics,15
Finance,15
NaN,5


In [87]:
# For Gender
df["Gender"].value_counts(dropna = False)

,count
Gender,
Male,72
Female,66
Other,7
NaN,5


In [88]:
# For City
df["City"].value_counts(dropna = False)

,count
City,
Pune,23
Chandigarh,22
Jaipur,17
Hyderabad,14
Delhi,14
Srinagar,13
Mumbai,13
Bengaluru,11
Chennai,9


In [89]:
# For Work_Mode
df["Work_Mode"].value_counts(dropna = False)

,count
Work_Mode,
Remote,56
Hybrid,48
Office,44
NaN,2


# 9. Handling Missing Values

In [90]:
df[["Age","Annual_Salary","Experience_Years","Performance_Score"]].isnull().sum()

,0
Age,4
Annual_Salary,4
Experience_Years,3
Performance_Score,3


In [91]:
# Fill NAN Values of Age with median computation
df["Age"] = df["Age"].fillna(df["Age"].median())

/tmp/ipykernel_3810/1470991933.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Age"] = df["Age"].fillna(df["Age"].median())


In [92]:
df["Age"].isnull().sum()

np.int64(0)

In [93]:
# For Annual Salary
df["Annual_Salary"] = df["Annual_Salary"].fillna(df["Annual_Salary"].median())

/tmp/ipykernel_3810/2188636908.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Annual_Salary"] = df["Annual_Salary"].fillna(df["Annual_Salary"].median())


In [94]:
df["Annual_Salary"].isnull().sum()

np.int64(0)

In [95]:
# Experience Years
df["Experience_Years"] = df["Experience_Years"].fillna(df["Experience_Years"].median())

/tmp/ipykernel_3810/2718623508.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Experience_Years"] = df["Experience_Years"].fillna(df["Experience_Years"].median())


In [96]:
df["Experience_Years"].isnull().sum()

np.int64(0)

In [97]:
# Performance Score
df["Performance_Score"] = df["Performance_Score"].fillna(
    df["Performance_Score"].median()
)

/tmp/ipykernel_3810/700465276.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Performance_Score"] = df["Performance_Score"].fillna(


In [98]:
df["Performance_Score"].isnull().sum()

np.int64(0)

# 10. Handling Missing Categorical Values

In [100]:
df[["Department","Gender","City","Work_Mode"]].isnull().sum()

,0
Department,5
Gender,5
City,5
Work_Mode,2


In [101]:
# Department
df["Department"] = df["Department"].fillna(
    df["Department"].mode()[0]
)

/tmp/ipykernel_3810/2759262574.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Department"] = df["Department"].fillna(


In [102]:
# Gender
df["Gender"] = df["Gender"].fillna(
    df["Gender"].mode()[0]
)

/tmp/ipykernel_3810/816209476.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Gender"] = df["Gender"].fillna(


In [103]:
# City
df["City"] = df["City"].fillna(
    df["City"].mode()[0]
)

/tmp/ipykernel_3810/3735380632.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["City"] = df["City"].fillna(


In [104]:
# Work Mode
df["Work_Mode"] = df["Work_Mode"].fillna(
    df["Work_Mode"].mode()[0]
)

/tmp/ipykernel_3810/2966138159.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Work_Mode"] = df["Work_Mode"].fillna(


In [106]:
df[["Department","Gender","City","Work_Mode"]].isnull().sum()

,0
Department,0
Gender,0
City,0
Work_Mode,0


# 11. Convert Date Column to DateTime

In [107]:
df["Joining_Date"].dtype

dtype('O')

In [108]:
df["Joining_Date"] = pd.to_datetime(
    df["Joining_Date"],
    errors="coerce"
)

/tmp/ipykernel_3810/3137497085.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Joining_Date"] = pd.to_datetime(


In [109]:
# verify invalid Dates
df["Joining_Date"].isnull().sum()

np.int64(0)

In [110]:
df["Joining_Date"].min()

Timestamp('2017-01-04 00:00:00')

In [111]:
df["Joining_Date"].max()

Timestamp('2026-01-06 00:00:00')

# 12. Remove Duplicate Records

In [112]:
df = df.drop_duplicates()

In [113]:
df.duplicated().sum()

np.int64(0)

In [114]:
df.shape

(150, 12)

In [115]:
# After Removing Duplicates Verify Employee IDs
df["Employee_ID"].duplicated().sum()

np.int64(0)

# 13. Verify Missing Values After Cleaning

In [116]:
# Verify Missing Values After Cleaning
df.isnull().sum().sum()

np.int64(0)

# 14. Verify Data Types After Cleaning

In [117]:
df.dtypes

,0
Employee_ID,object
Employee_Name,object
Department,object
Job_Title,object
Age,float64
Gender,object
Annual_Salary,float64
Experience_Years,float64
Joining_Date,datetime64[ns]
City,object


# 15. Verify Inconsistent Entries

In [118]:
# Now here is no longer Variations
df["Department"].unique()


array(['Engineering', 'Marketing', 'Human Resources', 'Sales', 'Finance',
       'Data & Analytics', 'Customer Success', 'Operations'], dtype=object)

In [119]:
df["Gender"].unique()

array(['Other', 'Female', 'Male'], dtype=object)

In [120]:
df["City"].unique()

array(['Pune', 'Mumbai', 'Bengaluru', 'Jaipur', 'Hyderabad', 'Chennai',
       'Chandigarh', 'Srinagar', 'Kolkata', 'Delhi'], dtype=object)

In [121]:
df["Work_Mode"].unique()

array(['Office', 'Remote', 'Hybrid'], dtype=object)

# 16. Check Numerical Data

In [122]:
# Validate Numerical Columns
df[["Age", "Annual_Salary", "Experience_Years", "Performance_Score"]].describe()

,Age,Annual_Salary,Experience_Years,Performance_Score
count,150.000000,150.000000,150.000000,150.000000
mean,38.113333,88854.833333,8.916667,3.560000
std,9.245740,22354.618782,4.929658,0.923082
min,21.000000,43331.000000,0.600000,2.000000
25%,31.000000,71123.500000,4.450000,3.000000
50%,39.000000,89302.000000,9.200000,4.000000
75%,46.000000,104693.250000,13.175000,4.000000
max,52.000000,162942.000000,17.400000,5.000000


In [123]:
df[(df["Age"] < 18) | (df["Age"] > 65)]

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode


In [124]:
df[
    (df["Performance_Score"] < 1) |
    (df["Performance_Score"] > 5)
]

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode


In [125]:
df[df["Experience_Years"] < 0]

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode


# 17. Save the original Dataframe

In [126]:
original_df = pd.read_csv("Messy_Company_Employee_Dataset.csv")

# 18. Comparison

In [127]:
comparison = pd.DataFrame({
    "Metric": [
        "Number of Rows",
        "Number of Columns",
        "Total Missing Values",
        "Duplicate Rows"
    ],
    "Before Cleaning": [
        original_df.shape[0],
        original_df.shape[1],
        original_df.isnull().sum().sum(),
        original_df.duplicated().sum()
    ],
    "After Cleaning": [
        df.shape[0],
        df.shape[1],
        df.isnull().sum().sum(),
        df.duplicated().sum()
    ]
})

comparison

,Metric,Before Cleaning,After Cleaning
0,Number of Rows,157,150
1,Number of Columns,12,12
2,Total Missing Values,32,0
3,Duplicate Rows,7,0


# 19. Final Dataset Verification

In [128]:
df.head()

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,Engineering,Data Scientist,48.0,Other,108371.0,13.2,2017-06-13,Pune,5.0,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.1,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28.0,Female,119400.0,13.2,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,9.2,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.9,2018-07-24,Hyderabad,4.0,Hybrid


In [129]:
df.tail()

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
152,EMP0132,Manya Rao,Marketing,Content Strategist,30.0,Female,49522.0,0.6,2021-11-25,Delhi,2.0,Remote
153,EMP0073,Karan Kapoor,Customer Success,Customer Success Manager,42.0,Male,43331.0,1.1,2018-03-23,Jaipur,3.0,Hybrid
154,EMP0124,Saira Kapoor,Engineering,QA Engineer,33.0,Male,94812.0,7.2,2024-06-12,Jaipur,4.0,Hybrid
155,EMP0020,Sara Ali,Sales,Sales Executive,22.0,Female,51745.0,2.4,2020-11-19,Pune,2.0,Office
156,EMP0120,Saira Sheikh,Marketing,Marketing Executive,52.0,Male,107328.0,17.4,2020-11-25,Bengaluru,2.0,Remote


In [130]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 150 entries, 0 to 156
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Employee_ID        150 non-null    object        
 1   Employee_Name      150 non-null    object        
 2   Department         150 non-null    object        
 3   Job_Title          150 non-null    object        
 4   Age                150 non-null    float64       
 5   Gender             150 non-null    object        
 6   Annual_Salary      150 non-null    float64       
 7   Experience_Years   150 non-null    float64       
 8   Joining_Date       150 non-null    datetime64[ns]
 9   City               150 non-null    object        
 10  Performance_Score  150 non-null    float64       
 11  Work_Mode          150 non-null    object        
dtypes: datetime64[ns](1), float64(4), object(7)
memory usage: 19.3+ KB


In [132]:
df.describe()

,Age,Annual_Salary,Experience_Years,Joining_Date,Performance_Score
count,150.000000,150.000000,150.000000,150,150.000000
mean,38.113333,88854.833333,8.916667,2021-04-06 15:12:00,3.560000
min,21.000000,43331.000000,0.600000,2017-01-04 00:00:00,2.000000
25%,31.000000,71123.500000,4.450000,2018-12-23 00:00:00,3.000000
50%,39.000000,89302.000000,9.200000,2021-01-05 00:00:00,4.000000
75%,46.000000,104693.250000,13.175000,2023-08-15 18:00:00,4.000000
max,52.000000,162942.000000,17.400000,2026-01-06 00:00:00,5.000000
std,9.245740,22354.618782,4.929658,NaN,0.923082


In [133]:
print("Shape:", df.shape)
print("Missing values:", df.isnull().sum().sum())
print("Duplicate rows:", df.duplicated().sum())

Shape: (150, 12)
Missing values: 0
Duplicate rows: 0


# 20.Export Cleaned Dataset

In [134]:
df.to_csv("Cleaned_Company_Employee_Dataset.csv", index=False)

In [135]:
import os

os.path.exists("Cleaned_Company_Employee_Dataset.csv")

True

In [136]:
cleaned_df = pd.read_csv(
    "Cleaned_Company_Employee_Dataset.csv"
)

cleaned_df.head()

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,Engineering,Data Scientist,48.0,Other,108371.0,13.2,2017-06-13,Pune,5.0,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.1,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28.0,Female,119400.0,13.2,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,9.2,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.9,2018-07-24,Hyderabad,4.0,Hybrid


# 21. Key observations
1. The original dataset
contained 157 records and 12 columns.
2. There were 27 missing values across several columns.
3. 7 duplicate records were identified and removed.
4. Categorical inconsistencies were found in Department, Gender, City, and Work Mode due to capitalization and spacing differences.
5. Numerical missing values were handled using median imputation, while categorical missing values were handled using mode imputation.
6. Joining_Date was converted from object/string format to proper datetime format.
7. After cleaning, the dataset contained 150 records, 12 columns, 0 missing values, and 0 duplicate rows.
8. The cleaned dataset was exported as Cleaned_Company_Employee_Dataset.csv.